File: Sample/UserpilotObjcSample/ObjcExample/Features/IdentifyViewController.m
//
//  SignInViewController.m
//  UserpilotObjcExample
//
//  Created by Motasem Hamed on 17/07/2025.
//

#import "IdentifyViewController.h"
#import "Userpilot+Shared.h"

@interface IdentifyViewController ()
@property (weak, nonatomic) IBOutlet UITextField *userIDTextField;
@property (weak, nonatomic) IBOutlet UITextField *userKeyTextField;
@property (weak, nonatomic) IBOutlet UITextField *userKeyValueTextField;

@property (weak, nonatomic) IBOutlet UITextField *companyIDTextField;
@property (weak, nonatomic) IBOutlet UITextField *companyKeyTextField;
@property (weak, nonatomic) IBOutlet UITextField *companyKeyValueTextField;

@end

@implementation IdentifyViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    [[Userpilot shared] screen:@"identify"];
}

- (IBAction)identifyTapped:(UIButton *)sender {
    if (_userIDTextField.text == nil || _userIDTextField.text.length == 0) return;
    NSString *userID = _userIDTextField.text;

    NSMutableDictionary<NSString *, NSString *> *properties = [NSMutableDictionary dictionary];
    if (_userKeyTextField.text != nil && _userKeyTextField.text.length != 0 &&
        _userKeyValueTextField.text != nil && _userKeyValueTextField.text.length != 0){
        properties[_userKeyTextField.text] = _userKeyValueTextField.text;
    }

    NSMutableDictionary<NSString *, NSString *> *company = [NSMutableDictionary dictionary];
    
    if (_companyIDTextField.text != nil && _companyIDTextField.text.length != 0) {
        NSString *companyID = _companyIDTextField.text;
        company[@"id"] = companyID;
        if (_companyKeyTextField.text != nil && _companyKeyTextField.text.length != 0 &&
            _companyKeyValueTextField.text != nil && _companyKeyValueTextField.text.length != 0){
            company[_companyKeyTextField.text] = _companyKeyValueTextField.text;
        }
    }

    [[Userpilot shared] identifyWithUserId:userID properties:properties  company:company];

    _userIDTextField.text = nil;
    _userKeyTextField.text = nil;
    _userKeyValueTextField.text = nil;
    _companyIDTextField.text = nil;
    _companyKeyTextField.text = nil;
    _companyKeyValueTextField.text = nil;
}

- (IBAction)signOutTapped:(UIStoryboardSegue *)unwindSegue {
    // Unwind to Sign Out
    [[Userpilot shared] logout];
}

- (IBAction)anonymousUserTapped:(UIButton *)sender {
    [[Userpilot shared] anonymous];
}

@end

File: Sample/UserPilotSample/UserPilotSample/Featues/Identify/IdentifyViewController.swift
//
//  IdentifyViewController.swift
//  UserpilotSample
//
//  Created by Motasem Hamed on 11/08/2024.
//

import Foundation
import UIKit

class IdentifyViewController: BaseViewController {

    // MARK: - IBOutlet

    @IBOutlet weak var textFieldUserID: UITextField! {
        didSet {
            textFieldUserID.delegate = self
        }
    }
    @IBOutlet weak var userPropertiesStackView: UIStackView!
    @IBOutlet weak var companyPropertiesStackView: UIStackView!
    @IBOutlet weak var anonymousButton: UIButton! {
        didSet {
            anonymousButton.layer.borderColor = UIColor.accent.cgColor
            anonymousButton.layer.borderWidth = 1
        }
    }

    internal var userPropertiesViews: [String: PropertyView] = [:]
    internal var companyPropertiesViews: [String: PropertyView] = [:]

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        UserpilotManager.shared.screen("identify")
    }

    // MARK: - IBActions

    @IBAction func onIdentifyButtonClicked(_ sender: UIButton) {
        identifyUser()
    }

    @IBAction func onBackButtonClicked(_ sender: UIButton) {
        close()
    }

    @IBAction func onAddUserProperty(_ sender: UIButton) {
        showAddUserPropertyDiaog()
    }

    @IBAction func onAddCompanyProperty(_ sender: UIButton) {
        showAddCompanyPropertyDialog()
    }

    @IBAction func onLogout(_ sender: UIButton) {
        FlowRoutingManager.shared.showAlertMessage("User logged out successfully!")
        UserpilotManager.shared.logout()
    }

    @IBAction func onAnonymous(_ sender: UIButton) {
        UserpilotManager.shared.anonymous()
    }

}

// MARK: - UITextFieldDelegate

extension IdentifyViewController: UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
}

// MARK: - Helper methods

extension IdentifyViewController {

    private func identifyUser() {
        guard
            let userID = textFieldUserID.text, !userID.isEmpty
        else {
            FlowRoutingManager.shared.showAlertMessage("Please insert User ID!")
            return
        }
        var userProperties: [String: String] = [:]
        for (_, propertyView) in userPropertiesViews {
            userProperties[propertyView.getParams().0] = propertyView.getParams().1
        }
        var companyProperties: [String: String] = [:]
        for (_, propertyView) in companyPropertiesViews {
            companyProperties[propertyView.getParams().0] = propertyView.getParams().1
        }
        UserpilotManager.shared.identify(userID: userID, properties: userProperties, company: companyProperties)
    }

}

// MARK: - Instance

extension IdentifyViewController {

    static func newInstance() -> IdentifyViewController {
        return IdentifyViewController()
    }

}
